File: python/extension2/VisitorWithLastChance.py
from __future__ import annotations
import random
from typing import TYPE_CHECKING

from python.root.Visitor import Visitor

if TYPE_CHECKING:
    from python.root.CoffeeHouse import CoffeeHouse
    from python.root.Product import Product


class VisitorWithLastChance(Visitor):
    def __init__(self, model: CoffeeHouse):
        super().__init__(model)

    def failDesiredProduct(self, currentTime: int, product: Product) -> None:
        if self._successRating == 5:
            self._successRating = 4
            menu = self._model.getMenu()
            self._desiredProduct = menu[random.choice(range(len(menu)))]
        else:
            super().failDesiredProduct(currentTime, product)

File: cplusplus/ext2/VisitorWithLastChance.cpp
#include "VisitorWithLastChance.hpp"

VisitorWithLastChance::VisitorWithLastChance(CoffeeHouse *model): Visitor(model)
{
}

void VisitorWithLastChance::failDesiredProduct(long currentTime, Product *product)
{
	if (this->successRating == 5)
	{
		this->successRating = 4;
		auto menu = this->model->getMenu();
		this->desiredProduct = menu[rand() % menu.size()];
	}
	else
	{
		Visitor::failDesiredProduct(currentTime, product);
	}
}